packages/core/src/auth/sso/ssoAccessTokenProvider.ts

```diff
@@ -651,9 +652,46 @@ class WebAuthorization extends SsoAccessTokenProvider {
         const cachedRegistration = await this.cache.registration.load(cacheKey)
 
         if (cachedRegistration && (isExpired(cachedRegistration) || cachedRegistration.flow !== 'web auth code')) {
-            await this.invalidate()
+            await this.invalidate('registrationExpired:WebAuth')
         }
 
         return loadOr(this.cache.registration, cacheKey, () => this.registerClient())
     }
 }
+
+/**
+ * Remembers the reason an SSO session was put in to a "needs reauthentication" state.
+ * The current use is for telemetry. When the user reauths, we want {@link AwsLoginWithBrowser}
+ * to know why it needed to be reauthed.
+ *
+ * The flow is to use `set()` to remember why the user was put in to a reauth state,
+ * then upon the next reauth use `get()`. Finally, use `clear()` if the reauth is
+ * successful.
+ */
+export class ReAuthState extends MemoryMap<ReAuthStateKey, ReAuthStateValue> {
+    static #instance: ReAuthState
+    static get instance() {
```

Comment:
since ReAuthState lives as `this.reauthState` on the sso objects, can we eliminate this singleton to avoid accidental usage
```suggestion
```

Reply:
Nope, since the SsoAccessTokenProvider instance is created at the moment of use in different flows: https://github.com/aws/aws-toolkit-vscode/blob/b0965b7b768c957d740511ec2fd76a658794807a/packages/core/src/auth/auth.ts#L701.

We could move ReAuthState in to `auth.ts`, but I'd like to avoid that for now to keep ReAuthState contained to a single file.